### ISSUE TITLE
Allow Class E IP addresses to better support Fan Networking

### ISSUE THREAD

**Issue by justinddunlap** [microsoft/WSL#4884]
**Is your feature request related to a problem? Please describe.**
Class E IPv4 network addresses are coming into broad use, particularly for container solutions, but Windows and thus WSL does not support this address space.

**Describe the solution you'd like**
Many container systems are now using what's called [Fan Networking](https://ubuntu.com/blog/introducing-the-fan-simpler-container-networking) - a more sophisticated form of overlay - to add additional IP address space, so that each IP address can have 250 addresses behind it, without having to maintain an overlay address database. While an official standard is still being worked on, solutions built on fan networking (particularly container solutions built on e.g. Docker) generally have a convention to use Class E addresses (suffixed with octet 240 and up) as an "expansion" class of IP address, meaning they assign these IP addresses to virtual bridges and other network devices.

Currently, Windows explicitly disallows Class E network addresses, meaning that attempting to assign a Class E address even under WSL results in an "Invalid argument" error message from RTNETLINK. This means that container solutions that expect these addresses to work (and work fine on e.g. Ubuntu and OS X) fail to work on WSL. This is the sole reason for me currently being unable to use Windows+WSL at work to develop via an in-house virtual application stack solution at my large organization. 

**Describe alternatives you've considered**
Fan networking *can* use IP addresses in other ranges, but this is discouraged, as you run a high risk of conflicting with existing IP addresses on the network. Because of this, many emerging solutions (such as the one I attempted to use) do not allow you to use addresses outside of the Class E space.

**Additional context**
Setting an IP address with a prefix of 240 on Ubuntu 18.04 on WSL 1 (Windows 10 1909):

```
$ sudo ip addr add '240.10.0.1/24' dev docker0
RTNETLINK answers: Invalid argument
```

Setting this IP address on the same bridge in Windows directly results in a message that only IP addresses prefixed with octets between 1 and 223 are valid.

**Comment by therealkenc** [comment]
> on WSL 1 

Repro seems to take on WSL2. Whether anything _other than_ the repro taking, couldn't tell you.

Were the Ubuntu kernel patches ([this???](https://bugs.launchpad.net/ubuntu/+source/ubuntu-fan/+bug/1470091)) ever upstreamed to mainline? Honest question, I can't tell. [Calling this tech 'fan' is most unfortunate. Defeats search entirely.]

![image](https://user-images.githubusercontent.com/7049700/74194350-ea745600-4c0d-11ea-88f6-83e35983e404.png)

**Comment by justinddunlap** [comment]
@therealkenc thanks for trying it! Do bridges created in WSL2 show up in Windows? Does Windows itself support configuring Class E IPs on devices now? Can it even communicate with or recognize a Class E IP?

As for search, "fan networking" in quotes works decently as a search term.

**Comment by therealkenc** [comment]
> Do bridges created in WSL2 show up in Windows? 

No. That would imply a win32 device driver.

I could not find an authoritative regarding whether fan is in the upstream kernel (nevermind Windows), or if it is, in what mainline kernel release. I stopped there. Hence, "Whether anything other than the repro taking, couldn't tell you" as the answer to your remaining queries. I tried nothing fan related. I tried your provided repro, which on WSL2 does not `EINVAL`. 

> Can it even communicate with or recognize a Class E IP?

Right now Windows doesn't bridge class-anything into WSL per #4150. The Hyper-V VM is behind a NAT.

> Also, I assume I need to install a pre-release version of Windows to use WSL2?

Yes. But that is not worth trying unless fan devices exist in the upstream kernel (specifically 4.19). 

> As for search, "fan networking" in quotes works decently as a search term.

It does not, or I wouldn't have asked you: "Were the Ubuntu kernel patches ever upstreamed to mainline?"

A quick kernel [search](https://livegrep.com/search/linux?q=fan&fold_case=auto&regex=false&context=true) suggests the negative, but I didn't want to make assumptions based on 20 second / three letter investigation when I looked yesterday. There does seem to be some traffic in the Canonical kernel [here](https://git.launchpad.net/~canonical-kernel/ubuntu/+source/ubuntu-fan/log/?h=master). One could imagine in the hypothetical those making their way to the mainline kernel, and then to the [MSFT Linux kernel](https://github.com/microsoft/WSL2-Linux-Kernel) at some point in the future.
